### ISSUE TITLE
How should we access capturing group content?

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/patterns#27]
Normally we just return the result as:
```js
var word = quantify(range("a","z"), 1, Infinity);
var number = quantify(digit(), 1, Infinity);
var wordAndNumber = concatenate(word, number);
var pattern = compile(wordAndNumber);
pattern.capture(number);
var string = "abc123";
var match = pattern.match(string);
```

Normally we just return an object like:

```js
{
	value: "abc123",
	pattern: wordAndNumber,
	subpatterns: [
		{
			value: "123",
			pattern: number
		}
	]
}
```

This is easy to process programmatically, but if we don't need all of the capturing groups or if we want to process it in a specific, not in a general way, then it is inconvenient.

A possible solution to reuse capturing group selection code:
```js
var value = match.select(number);
```

**Comment by inf3rno** [comment]
What I don't like here is this:

```js
pattern.capture(number);
match.select(number);
```

We have to create the same selector twice. What we really need here is defining the selector only once. A possible solution is adding a string identifier for the selector output.

```js
var word = quantify(range("a","z"), 1, Infinity);
var number = quantify(digit(), 1, Infinity);
var wordAndNumber = concatenate(word, number);
var pattern = compile(wordAndNumber);
pattern.capture(number).as("num");
var string = "abc123";
var match = pattern.match(string);
var num = match.get("num");
```

It is easier to reuse a string than rebuilding a complete selector. We lose some functionality here, because we could write a different capturing selector and a different output selector, but I think that is unlikely. It would happen only if we process the same pattern with the same capturing groups in multiple different ways. So one way needs one set of capturing groups, while the other way needs a different set of capturing groups. There can be overlaps, etc. It might be better this way, because we match the same pattern only once instead of compiling and matching 2 patterns just to have different capturing groups. On the other hand the 2 pattern approach can be scaled horizontally, while the 1 pattern approach means vertical scaling.

**Comment by inf3rno** [comment]
Another thought that we could use function calls or callback functions for our selector rules.

```js
var word = quantify(range("a","z"), 1, Infinity);
var number = quantify(digit(), 1, Infinity);
var wordAndNumber = concatenate(word, number);
var pattern = compile(wordAndNumber);
pattern.capture(number);
var string = "abc123";
var match = pattern.match(string);
match.fetch(function (number){
    // ...
});
```

I don't like this. I think I though of using the callback for building the output. So it should be added to the capture part instead.

```js
var word = quantify(range("a","z"), 1, Infinity);
var number = quantify(digit(), 1, Infinity);
var wordAndNumber = concatenate(word, number);
var pattern = compile(wordAndNumber);
pattern.capture(number, function (number, match){
    match.number = parseInt(number);
});
var string = "abc123";
var match = pattern.match(string);
console.log(match.number);
```

I think this approach is what I was talking about. We have to use the selector only once. We use only the values to build the output and we exactly know which usages we are talking about. We can even handle them as a collection. We can compile a pattern multiple times with different capturing groups if we want to.

**Comment by inf3rno** [comment]
We can still use the output format, but it would be an internal format and it will use the pattern ids instead of the pattern objects.

**Comment by inf3rno** [comment]
I think it is perfect. Now we have to decide about the selectors and the arguments of the callback functions and what to support by building the output.

**Comment by inf3rno** [comment]
So what is clear here that the compiled pattern with the capturing groups and the output converter is one closed unit. Its input is the string (or pointer) and its output is a normal js object. I am not sure about the pointer. I mean it can be an internal object too and we could do `pattern.match(string, fromIndex)` instead of `pattern.match(pointer(string, fromIndex))`. I need the pointer only because we can use it for iteration. Actually in parsers we use a single pointer with multiple patterns, so we need to export them. What is interesting about the capturing group callbacks that we don't necessarily use them for building an output. I think we'll need a second "call only" method. So the `pattern.match` will create an output object, while this `pattern.apply` or `pattern.process` will call the capturing callbacks, but will not return anything.

**Comment by inf3rno** [comment]
I think the compile is not the proper word here, because `var pattern = compile(wordAndNumber);`  means that we use the `wordAndNumber` as base, but the compilation comes only after we added the capturing groups too.

The whole process is something like
```js
use(wordAndNumber)
    .capture(number, fn)
    .compile()
    .match(string)
```

**Comment by inf3rno** [comment]
We can do lazy compilation if we compile by first match, so we can skip that part on the API.

**Comment by inf3rno** [comment]
I am not sure which words to use. We could skip even the `use` and do `wordAndNumber.capture(number, fn)` for lazy pattern creation, but in that case the capture return value must be the pattern object instead of the capturing group object, which can be confusing. Another issue, that we cannot configure the capturing group in multiple calls that way, so I'd rather avoid that solution if possible.

**Comment by inf3rno** [comment]
Let's continue this:
```js
pattern.capture(number, function (number, match){
    match.number = parseInt(number);
});
```

Ok. Actually we will never have a subpattern called number, but we can neglect that part now. What is important here, that the callback function is too complicated. We must have something like:

```js
pattern.capture(number, parseInt);
```

It cannot be more complicated than that. We need to find a different way than passing parameters to communicate the rest of the info. Actually in most of the case the subpattern match should not know about the whole pattern match and the result object we want to build for that. So this cb is only responsible for the string - number conversion and the higher level conversion should care about how to add it to a result object.

**Comment by inf3rno** [comment]
```js
var pattern = use(wordAndNumber);
pattern.capture(number, parseInt);
var string = "abc123";
var match = pattern.match(string);
console.log(match); // [123]
```

I think it should return an array of the results normally if we don't give anything to hand le the highest level match (the whole string in this case). We could do

```js
var pattern = use(wordAndNumber);
pattern.capture(number, parseInt);
pattern.capture(wordAndNumber, [number], function (string, [number]){
    return number;
});
var string = "abc123";
var match = pattern.match(string);
console.log(match); // 123
```

But it looks ugly, and we don't really want to capture the whole string in this case. I'd rather add some sort of non capturing main transformation.

```js
var pattern = use(wordAndNumber);
pattern.capture(number, parseInt);
pattern.compile(number, function (number){
    return number;
});
var string = "abc123";
var match = pattern.match(string);
console.log(match); // 123
```
We can use the compile for something after all. :P

**Comment by inf3rno** [comment]
Under the hood we never keep the reference to the  `number` or `wordAndNumber` objects here. They got and id, probably a dependency tree, and a capturing group mapping will be built with those ids, and after the compilation is done they will be freed.

**Comment by inf3rno** [comment]
I worked on this a little bit more and ended up with the following

```js
var word = quantify(range("a","z"), 1, Infinity);
var number = quantify(digit(), 1, Infinity);
var wordAndNumber = concatenate(word, number);
var numberCapturingGroup = capture(number, parseInt);
var wordAndNumberMatcher = compile(wordAndNumber, numberCapturingGroup, t.one);
var string = "abc123";
var match = wordAndNumberMatcher.findNext(string);
console.log(match.value); // 123
```
